core/jvm/src/main/scala/fs2/compress.scala

```diff
@@ -25,45 +22,74 @@ object compress {
       nowrap: Boolean = false,
       bufferSize: Int = 1024 * 32,
       strategy: Int = Deflater.DEFAULT_STRATEGY
-  ): Pipe[F, Byte, Byte] = { in =>
-    Pull.suspend {
-      val deflater = new Deflater(level, nowrap)
-      deflater.setStrategy(strategy)
-      val buffer = new Array[Byte](bufferSize)
-      _deflate_stream(deflater, buffer)(in)
-    }.stream
+  ): Pipe[F, Byte, Byte] = stream => {
+    val deflater = new Deflater(level, nowrap)
+    deflater.setStrategy(strategy)
+    _deflate(deflater, bufferSize, crc32 = None)(stream) ++
+      Stream.suspend {
+        deflater
+          .end() // If F is narrowed to Applicative then deflater.end() can be bracketed rather than appended in an empty stream.
```

Comment:
What happens if this never gets called? E.g., `src.through(compress.deflate).take(10)`

Reply:
Looks like openjdk allocates native memory for both `Inflater` / `Deflater` so we really need to bracket this somehow in order to ensure that native memory is released.

Reply:
Agreed. That's correct, not calling deflater.end() or inflater.end() risks (will) leaking native memory - an issue present in the existing implementation and why I'd like to constrain F to Applicative so that the deflate/inflate/gzip/gunzip streams can be bracketed, but this introduces a binary incompatibility.

Reply:
We'd need a full `F[_]: Sync` -- we need the `delay` constructor.

We could leave the old implementation in place and deprecate it and then define the new implementation to require the `Sync` constraint.

Reply:
Can it be done with a `Stream.bracket` instead?

Reply:
@djspiewak Yeah, it would be done via `Stream.bracket` -- the reason we need a `Sync[F]` is so we can lift creation (e.g. `Sync[F].delay(new Inflater)`) and disposal (e.g. `Sync[F].delay(inflater.end())`) in an effect. I assume the `Applicative[F]` suggestion was in order to get access to `pure` but that's not enough -- we'd need `delay`.

Reply:
Ahhh, I see what you mean.

Reply:
Agreed. fs2.compress and CompressSpec have been restored to their original implementations, for the necessary type changes to be implemented in fs2.compress2 - naming opinions are welcome.

Reply:
Right, the functions fs2.compress2.(deflate|inflate|gzip|gunzip) are now bracketed.